llvm/test/CodeGen/AIE/aie2/GlobalISel/inst-select-combine-shift-sext.mir

```diff
@@ -0,0 +1,132 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py
+#
+# This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#
+# (c) Copyright 2023-2024 Advanced Micro Devices, Inc. or its affiliates
+# RUN: llc -mtriple aie2 -start-before=instruction-select -stop-after=dead-mi-elimination %s \
+# RUN:     -verify-machineinstrs -o - | FileCheck %s
+
+# In this test, both shifts can be combined with G_SEXT_INREG.
+
+---
+name:            combineBeforeAndAfter
+legalized:       true
+regBankSelected: true
+
+body: |
+  bb.0:
+    ; CHECK-LABEL: name: combineBeforeAndAfter
+    ; CHECK: [[COPY:%[0-9]+]]:ep = COPY $p0
+    ; CHECK-NEXT: [[LDA_dms_lda_idx_imm:%[0-9]+]]:er = LDA_dms_lda_idx_imm [[COPY]], 0 :: (load (s32))
+    ; CHECK-NEXT: [[MOV_RLC_imm10_pseudo:%[0-9]+]]:er = MOV_RLC_imm10_pseudo 16
+    ; CHECK-NEXT: [[LSHL:%[0-9]+]]:er = LSHL [[LDA_dms_lda_idx_imm]], [[MOV_RLC_imm10_pseudo]]
+    ; CHECK-NEXT: [[MOV_RLC_imm10_pseudo1:%[0-9]+]]:er = MOV_RLC_imm10_pseudo -16
+    ; CHECK-NEXT: [[ASHL:%[0-9]+]]:er = ASHL [[LSHL]], [[MOV_RLC_imm10_pseudo1]]
+    ; CHECK-NEXT: $r0 = COPY [[ASHL]]
+    ; CHECK-NEXT: PseudoRET implicit $lr, implicit $r0
+    %0:ptrregbank(p0) = COPY $p0
+    %1:gprregbank(s32) = G_LOAD %0 :: (load (s32), align 4)
+    %2:gprregbank(s32) = G_CONSTANT i32 4
+    %3:gprregbank(s32) = G_SHL %1:gprregbank(s32), %2:gprregbank(s32)
+    %4:gprregbank(s32) = G_SEXT_INREG %3:gprregbank(s32), 20
+    %5:gprregbank(s32) = G_ASHR %4:gprregbank(s32), %2:gprregbank(s32)
```

Comment:
Is this combine really specific to AIE?
Could this be written as a generic pre/post-legalization combiner instead?

Reply:
Hi @konstantinschwarz, thank you for the idea. I completely changed the implementation. I Solved the problem in legalization phase, in this way, all sifts will be combined through already available combiners.